opm/autodiff/BlackoilModelParametersEbos.hpp

```diff
@@ -78,6 +78,11 @@ SET_SCALAR_PROP(FlowModelParameters, MaxPressureChangeMsWells, 2.0 *1e5);
 SET_BOOL_PROP(FlowModelParameters, UseInnerIterationsMsWells, true);
 SET_INT_PROP(FlowModelParameters, MaxInnerIterMsWells, 10);
 
+// if openMP is available, determine the number threads per process automatically.
+#if _OPENMP
+SET_INT_PROP(FlowModelParameters, ThreadsPerProcess, -1);
+#endif
+
```

Comment:
Is this the substitute for the change in /OPM/ewoms#337 [fvbasediscretization.hh#L212](/OPM/ewoms/blob/6e3891554af7cc6ef2fed4c728c083e21120b32d/ewoms/disc/common/fvbasediscretization.hh#L212):
```
@@ -209,7 +209,7 @@ SET_TYPE_PROP(FvBaseDiscretization, ConstraintsContext, Ewoms::FvBaseConstraints
   * \brief The OpenMP threads manager
   */
  SET_TYPE_PROP(FvBaseDiscretization, ThreadManager, Ewoms::ThreadManager<TypeTag>);
- SET_INT_PROP(FvBaseDiscretization, ThreadsPerProcess, -1);
+ SET_INT_PROP(FvBaseDiscretization, ThreadsPerProcess, 1);
   SET_BOOL_PROP(FvBaseDiscretization, UseLinearizationLock, true);
/*!
```

This change from FvBaseDiscretization to FlowModelParameters really confuses me. Is the first parameter irrelevant or is there some inheritance graph between the two that I do not see.

Reply:
> Is this the substitute for the change in /OPM/ewoms#337

in some sense yes: this makes sure the threading behavior of `flow` does not change regardless of whether OPM/ewoms#377 is merged or not. OPM/ewoms#377 fixes a race condition caused by the fact that `Dune::DynamicMatrix::resize()` is not threadsafe and reverts the eWoms default threading behavior to that of before  OPM/ewoms#370 (i.e., threading must be enabled explicitly on the command line), so in some sense these two PRs are orthogonal.

> his change from FvBaseDiscretization to FlowModelParameters really confuses me. Is the first parameter irrelevant or is there some inheritance graph between the two that I do not see.

for `flow` properties for the latter take preceedence those of the former: The FvBaseDiscretization is the ancestor of all type tags used by any eWoms based simulator.

you can get the inheritance hierarchy by

```
flow --print-properties=1 | grep -i "inherited from"
```

if you put this through image processing and merge duplicate edges, you'll get a nice looking graph.